### ISSUE TITLE
Manual backpropagation

### ISSUE THREAD

**Issue by aespielberg** [taichi-dev/taichi#267]
Sorry, I forgot I had one more question.  I would like to do longer simulations than 1024 steps.  This will require manual backpropagation, because I don't want to overload the GPU.  This leads me to two questions:

1. Is there a preferred way to manually backprop?  It seems the preferred way to do backprop is not to explicitly write, or even call a backward function, but to use compute_loss.
2. If there is a stable way to manually backprop, is there a way to pass in the entire gradient of the state, and a convenient way to capture that?  Similar to grad_ys in tf.gradients or the grad argument in pytorch?

**Comment by yuanming-hu** [comment]
You can always invoke `p2g.grad()` for the gradient version of `p2g`

**Comment by aespielberg** [comment]
Hmm.  But if I do n forward steps and collect n "memos," as we did before, then I need to do backprop of every step backward.  and I'll need to carry along the "future" gradient with me, and feed it into p2g's gradient version, somehow.

Is this the same as intiializing the tensors' .grad with the .grad I already computed from the "next" timestep?

**Comment by yuanming-hu** [comment]
Sorry for missing this message. Yes, you will have to copy the gradients into tensors such as x.grad and invoke p2g. The role of the gradients kernels is to take as input the gradients of its output tensors and accumulate the input tensor gradients.

**Comment by aespielberg** [comment]
Thanks.  I'll try to do a few extremely simple tests to make sure I understand this.  One other question - how do I know when I'm using up the entire GPU memory and spilling over (aside from nvidia-smi).  Does/can the compiler give a warning?
